### PR TITLE
nixos/goss: init

### PR DESCRIPTION
**Pull request by anthonyroussel**
## Description of changes

Add Goss NixOS module which starts a Goss SystemD service serving a /healthz endpoint to validate the server configuration. See https://goss.rocks

Some possible improvements when I'll have time: add support for gossfile templating, add systemd hardening

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by anthonyroussel** [comment]
@ofborg test goss

**Review by SuperSandro2000 — APPROVED** [review]

File: nixos/modules/services/monitoring/goss.nix [hunk 1]
```diff
@@ -0,0 +1,73 @@
+{ config, lib, pkgs, ... }:
+
+let
+  cfg = config.services.goss;
+
+  settingsFormat = pkgs.formats.yaml { };
+  configFile = settingsFormat.generate "goss.yaml" cfg.settings;
+
+in {
+  meta.maintainers = [ lib.maintainers.anthonyroussel ];
```
Comment (pbsds) [inline nixos/modules/services/monitoring/goss.nix:10]: Please add `meta.doc`

File: nixos/tests/goss.nix [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+
+    machine.wait_for_unit("goss.service")
+    machine.wait_for_open_port(8080)
+
+    with subtest("returns health status"):
+      result = json.loads(machine.succeed("curl -sS http://localhost:8080/healthz"))
+
+      assert len(result["results"]) == 10, ".results should be an array of 10 items"
+      assert result["summary"]["failed-count"] == 0, ".summary.failed-count should be zero"
+      assert result["summary"]["test-count"] == 10, ".summary.test-count should be 10"
```
Comment (pbsds) [inline nixos/tests/goss.nix:77]:
```suggestion
      assert len(result["results"]) == 10, f".results should be an array of 10 items, was {result['results']!r}"
      assert result["summary"]["failed-count"] == 0, f".summary.failed-count should be zero, was {result['summary']['failed-count']}"
      assert result["summary"]["test-count"] == 10, f".summary.test-count should be 10, was {result['summary']['test-count']}"
```

File: pkgs/tools/misc/goss/default.nix [hunk 3]
```diff
@@ -26,17 +32,30 @@ buildGoModule rec {
...
+
   checkFlags = [
     # Prometheus tests are skipped upstream
     # See https://github.com/goss-org/goss/blob/master/ci/go-test.sh
     "-skip" "^TestPrometheus"
   ];
 
+  postInstall = let
+    runtimeDependencies = [ bash getent ]
+      ++ lib.optionals (!stdenv.isDarwin) [ systemd ];
```
Comment (pbsds) [inline pkgs/tools/misc/goss/default.nix:45]: Will be helpful once the BSDs get fully up and running

```suggestion
      ++ lib.optionals stdenv.isLinux [ systemd ];
```

**Comment by anthonyroussel** [comment]
@pbsds Hello. Thanks for your review! I have handled all your feedback.

**Review by pbsds — APPROVED** [review]
LGTM, waiting for ofborg on aarch64-darwin

Result of `nixpkgs-review pr 263335` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
    <li>tests.nixos-functions.nixos-test</li>
  </ul>
</details>
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>dgoss</li>
    <li>goss</li>
  </ul>
</details>

I've in addition built `goss.passthru.tests` and `.#htmlDocs.nixosManual`, and inspected the manual.

# package

- [x] package name fits guidelines
- [x] package version fits guidelines
- [x] package build on x86_64-linux
- [x] executables tested on x86_64-linux
- [x] all depending packages build
- [x] patches have a comment describing either the upstream URL or a reason why the patch wasn't upstreamed
- [x] patches that are remotely available are fetched rather than vendored

# module


- [x] module path fits the guidelines
- [x] module tests succeed on x86_64-linux
- [x] options have appropriate types
- [x] options have default
- [x] options have example
- [x] options have descriptions
- [x] No unneeded package is added to environment.systemPackages
- [x] meta.maintainers is set
- [x] module documentation is declared in meta.doc

**Pull request merged**